### ISSUE TITLE
Containers of different projects being merged despite renaming the containers

### ISSUE THREAD

**Issue by akaibukai** [docker/compose#7384]
Hi everybody,

Today I just noticed a very weird behavior regarding the `container_name` option when I have multiple project that share a common directory structure.

For example let's say I have the following two projects:
```
a_project
├─ README.md
├─ a.txt
└─ docker
   ├─ Dockerfile
   └─ docker-compose.yml

b_project
├─ README.md
├─ b.txt
└─ docker
   ├─ Dockerfile
   └─ docker-compose.yml
```
with the following docker-compose file for both project:
```
version: "3"
services:
  app:
    build:
      context: ..
      dockerfile: ./docker/Dockerfile
    volumes:
      - ../:/app
    stdin_open: true
    tty: true
```
The thing to notice here is that since the services have the same name (`app`) and that the relative parent folder have also the same name (`docker`), the resulting container upon a `docker-compose up` will have the following name: `docker_app_1`.

And the following is happening:
```
~/a_project/docker$ docker-compose up -d
Creating docker_app_1 ... done
~/a_project/docker$ cd ~/b_project/docker
~/b_project/docker$ docker-compose up -d
Recreating docker_app_1 ... done
```
This might be the expected behavior since the absolute directory containing the project doesn't really matter.

Now the weird thing is that if I explicitly set a different container name for each service, like the following:
```
version: "3"
services:
  app:
    build:
      context: ..
      dockerfile: ./docker/Dockerfile
    container_name: a_app            # inside the a_project
    # container_name: b_app          # inside the b_project 
    volumes:
      - ../:/app
    stdin_open: true
    tty: true
```
It appears that because of the service name (`app`) and the relative parent directory (`docker`) being still the same, this weird merging behavior is still happening!

To better see, here I'm not detaching the container with `-d` and I'm using two terminal windows:
```
# In one terminal window
~/a_project/docker$ docker-compose up
Creating a_app ... done
Attaching to a_app

a_app exited with code 137 #happening after running below command
```
```
# In another terminal window
~/b_project/docker$ docker-compose up -d
Recreating a_app ... done
Attaching to b_app
```
Here two things to notice:

- the container name displayed when recreating is the previous one currently running (`a_app`) despite I'm running the `b_project`, but then it correctly handle that using the correct name when attaching
- This automatically exited the previous container, hence the fact there is only one container being merged

For completeness if anyone want to reproduce this, here is the very simple Dockerfile I'm using:
```
FROM alpine

RUN mkdir -p /app
WORKDIR /app
```
_I kept the volumes things with an `/app` folder just to be sure to replicate my current projects structure that handle some volumes, but I thing there are not relevant regarding the issue.._

Now if I have either different names for the parent docker directory, e.g:
```
a_project
├─ README.md
├─ a.txt
└─ docker_a                 #different
   ├─ Dockerfile
   └─ docker-compose.yml

b_project
├─ README.md
├─ b.txt
└─ docker_b                 #different
   ├─ Dockerfile
   └─ docker-compose.yml
```
Or if I have different name for the services, e.g.:
```
version: "3"
services:
  a_app:                    #different
    build:
...

version: "3"
services:
  b_app:                    #different
    build:
...
```
**So in conclusion, how the `container_name` option is supposed to work?**

I'm on Linux 5.6.5 (manjaro/arch) and I have the following versions:
- Docker version 19.03.8-ce
- docker-compose version 1.25.5

**Comment by akaibukai** [comment]
Hi everybody,
I found that this behavior doesn't apply when I set a project name: https://docs.docker.com/compose/reference/overview/#use--p-to-specify-a-project-name

But I'm not sure if without this option, the way it's currently working is still the correct one.
Since I think that the containers names should have precedence..
